### ISSUE TITLE
FP16 quantization does not seem to work well

### ISSUE THREAD

**Issue by SM1991CODES** [NVIDIA-AI-IOT/torch2trt#520]
Hi,
I have a custom model working on bird's eye view of the KITTI dataset.
I tried setting the fp16_mode flag to true to get a 16 bit quantized model.
However, I find this:
1) Model size increases - original model 56MB, after quantization 70MBs.
2) I get lot's of nan for box rotation predictions.

I would like to add that everything works well without quantization.

Also, I plan to use int8 quantization as well. I found some posts about how to use them but still have some questions:
1) What about calibration? Like, how many samples do we need for calibration? Could you pleas provide a full example?
2) Do the calibration samples need to come from the validation set?

Please see the attached code section for my initial problem with fp16.

Best Regards
Sambit

`# net1.load_state_dict(torch.load(trained_model_path, map_location="cuda:0"))
    net1.load_state_dict(torch.load(trained_model_path, map_location="cuda:0"))
    net1.eval()  # set network to evaluation

    # convert to tensorrt if needed
    if use_trt:
        # no quantization
        if quant_bits == 32:
            net1 = torch2trt(net1, [torch.randn(1, 1, 512, 256).cuda()])  # get tensorrt model
        elif quant_bits == 16:
            net1 = torch2trt(net1, [torch.randn(1, 1, 512, 256).cuda()], fp16_mode=True)
            torch.save(net1.state_dict(), "net_fp16.pth")  # save to check size later

    # torchsummary.summary(net1, (settings.N_CHANNELS_INPUT_BEV, 512, 256))  # print out a summary

    # load the test dataset
    evalset = dataset_classes.ValidateBirdEyeViewDataset(path_validation_velodyne_bev=settings.validate_velodyne_bev_path_docker,
                                                         path_validation_velodyne_3d=settings.validate_velodyne_3d_path_docker)

    # define a data loader with batch size 1 since real sensor does not produce batches of data
    loader = DataLoader(evalset, batch_size=1)

    # to avoid CUDA memory errors during inference
    with torch.no_grad():
        frame_counter = 0
        msTimer1 = timer.msTimer()  # start timing the inference
        msTimer1.start()
        for index, batch in enumerate(loader):
            frame_counter += 1

            # get the normalized input range image, original range image, point cloud and frame number
            test_x, pcl, frame_num = batch[0],\
                                     np.squeeze(batch[1].numpy(), axis=0),\
                                     batch[2].numpy().item()

            test_x_copy = np.squeeze(test_x.numpy().copy(), axis=0)  # normalized range image
            # test_x_copy = test_x_copy * 255.  # multiple 255 to remove the normalization done
            test_x = test_x.permute(0, 3, 1, 2)
            test_x = test_x.cuda()

            # test_x = test_x.cpu()  # for qunatization, use CPU

            msTimer2 = timer.msTimer()  # start timing the inference
            msTimer2.start()

            # predictions
            predictions, boxes, roty = net1(test_x)

            delay_s, delay_ms = msTimer2.get_time_delta()
            print("average inference time per frame : ", delay_ms)`

**Comment by SM1991CODES** [comment]
Can anyone please help??

**Comment by jaybdub** [comment]
Hi @SM1991CODES ,

Thanks for reaching out!

Regarding FP16 conversion, are you able to share the model you're attempting to convert?  It's difficult to assess what might be happening here.

Regarding INT8 conversion.  The calibration dataset is simply used to compute dynamic ranges for the quantization.  So, you should use whatever data you expect to use in practice.  Training data should work fine.  I've heard ballpark estimates of ~500 images, but you might be able to get away with fewer, you would just have to test.  For the steps on how to perform INT8 calibration with torch2trt, I recommend following here

https://nvidia-ai-iot.github.io/torch2trt/v0.2.0/usage/reduced_precision.html

Please let me know if this helps or you have any questions.

Best,
John

**Comment by SM1991CODES** [comment]
Hello John,

Thank you for your reply. The network is basically a U-Net modified for key point detection.

`import torch
from torch import nn
import torch.nn.functional as F


class UNet(nn.Module):
    def __init__(
        self,
        in_channels=1,
        n_classes=2,
        depth=5,
        wf=6,
        padding=False,
        batch_norm=False,
        up_mode='upconv',
    ):
        """
        Implementation of
        U-Net: Convolutional Networks for Biomedical Image Segmentation
        (Ronneberger et al., 2015)
        https://arxiv.org/abs/1505.04597
        Using the default arguments will yield the exact version used
        in the original paper
        Args:
            in_channels (int): number of input channels
            n_classes (int): number of output channels
            depth (int): depth of the network
            wf (int): number of filters in the first layer is 2**wf
            padding (bool): if True, apply padding such that the input shape
                            is the same as the output.
                            This may introduce artifacts
            batch_norm (bool): Use BatchNorm after layers with an
                               activation function
            up_mode (str): one of 'upconv' or 'upsample'.
                           'upconv' will use transposed convolutions for
                           learned upsampling.
                           'upsample' will use bilinear upsampling.
        """
        super(UNet, self).__init__()
        assert up_mode in ('upconv', 'upsample')
        self.padding = padding
        self.depth = depth
        prev_channels = in_channels
        self.down_path = nn.ModuleList()
        for i in range(depth):
            self.down_path.append(
                UNetConvBlock(prev_channels, 2 ** (wf + i), padding, batch_norm)
            )
            prev_channels = 2 ** (wf + i)

        self.up_path = nn.ModuleList()
        for i in reversed(range(depth - 1)):
            self.up_path.append(
                UNetUpBlock(prev_channels, 2 ** (wf + i), up_mode, padding, batch_norm)
            )
            prev_channels = 2 ** (wf + i)

        self.last = nn.Conv2d(prev_channels, n_classes, kernel_size=1)

    def forward(self, x):
        blocks = []
        for i, down in enumerate(self.down_path):
            x = down(x)
            if i != len(self.down_path) - 1:
                blocks.append(x)
                x = F.max_pool2d(x, 2)

        for i, up in enumerate(self.up_path):
            x = up(x, blocks[-i - 1])

        # return self.last(x)
        # modified to return features
        return x


class UNetConvBlock(nn.Module):
    def __init__(self, in_size, out_size, padding, batch_norm):
        super(UNetConvBlock, self).__init__()
        block = []

        block.append(nn.Conv2d(in_size, out_size, kernel_size=3, padding=int(padding)))
        block.append(nn.ReLU())
        if batch_norm:
            block.append(nn.BatchNorm2d(out_size))

        block.append(nn.Conv2d(out_size, out_size, kernel_size=3, padding=int(padding)))
        block.append(nn.ReLU())
        if batch_norm:
            block.append(nn.BatchNorm2d(out_size))

        self.block = nn.Sequential(*block)

    def forward(self, x):
        out = self.block(x)
        return out


class UNetUpBlock(nn.Module):
    def __init__(self, in_size, out_size, up_mode, padding, batch_norm):
        super(UNetUpBlock, self).__init__()
        if up_mode == 'upconv':
            self.up = nn.ConvTranspose2d(in_size, out_size, kernel_size=2, stride=2)
        elif up_mode == 'upsample':
            self.up = nn.Sequential(
                nn.Upsample(mode='bilinear', scale_factor=2),
                nn.Conv2d(in_size, out_size, kernel_size=1),
            )

        self.conv_block = UNetConvBlock(in_size, out_size, padding, batch_norm)

    def center_crop(self, layer, target_size):
        _, _, layer_height, layer_width = layer.size()
        diff_y = (layer_height - target_size[0]) // 2
        diff_x = (layer_width - target_size[1]) // 2
        return layer[
            :, :, diff_y : (diff_y + target_size[0]), diff_x : (diff_x + target_size[1])
        ]

    def forward(self, x, bridge):
        up = self.up(x)
        crop1 = self.center_crop(bridge, up.shape[2:])
        out = torch.cat([up, crop1], 1)
        out = self.conv_block(out)

        return out`

Layers used:
- Conv2D, ConvTransposed2D, ReLU, cat, BatchNorm2D

While converting FP32 to tensorrt, everything works well. But with FP16, lots of input frames are with all 0 predicted masks and the few that actually have some prediction results have a lot of 'nan' values in them.
Also, I do not see any reduction in network size.

Best Regards
Sambit

**Comment by SM1991CODES** [comment]
Sorry for the bad code paste.
Please see the attached file.

Best Regards
Sambit
[code.txt](https://github.com/NVIDIA-AI-IOT/torch2trt/files/6082414/code.txt)
